openmc/data/multipole.py

```diff
@@ -531,7 +531,8 @@ def _evaluate(self, E, T):
         sqrtkT = sqrt(K_BOLTZMANN * T)
         sqrtE = sqrt(E)
         invE = 1.0 / E
-        dopp = self.sqrtAWR / sqrtkT
+        if sqrtkT > 0.0:
+            dopp = self.sqrtAWR / sqrtkT
```

Comment:
That is correct; I added the `if` statement to avoid the divide by zero warning which might worry a user.

Reply:
Oh, I never noticed before that standard Python floats will raise a `ZeroDivisionError`.  That's kind of annoying.  I wonder why the Python devs decided to break IEEE compliance.  In practice, it's unlikely a user will see that error because `self.sqrtAWR` is usually loaded from HDF5 as type `numpy.float64` which gives the obviously superior behavior of producing an inf rather than raising an exception.

Reply:
IEEE 754 defines five standard floating point exceptions, of which divide-by-zero is one. In languages that have exceptions, like Python, raising an exception is the right thing to do. Normally that is what happens -- go ahead and try 1.0/0.0 in an interactive prompt. NumPy is actually what allows the divide-by-zero to proceed without an exception; instead, it just shows a warning.

Reply:
Do I have something configured strangely?  Numpy floats do not show a divide-by-zero warning for me.

I interpret IEEE 754 differently.  See [this FAQ](http://grouper.ieee.org/groups/754/faq.html#exceptions) which talks about this in plain English.  The 754-2008 standard says this in section 7.1:

> This clause also specifies default non-stop exception handling for exception signals, which is to deliver a default result, continue execution, and raise the corresponding status flag

There's no way to continue execution in Python with the IEEE default result of inf without something silly like putting try-except around every division operation:
```python
    try:
        c = a / b
    except ZeroDivisionError:
        c = float('inf')
```

Reply:
Here's what I get:
```Python
>>> import numpy as np
>>> one = np.float_(1.0)
>>> zero = np.float_(0.0)
>>> one/zero
/home/romano/.pyenv/versions/3.6.2/bin/ptpython:1: RuntimeWarning: divide by zero encountered in double_scalars
  #!/home/romano/.pyenv/versions/3.6.2/bin/python3.6
inf
```

As to why Python raises an exception, it is in line with the overarching philosophy of Python to be simple, understandable, and predictable. In fact, the Zen of Python explicitly says "Errors should never pass silently. Unless explicitly silenced." If divide-by-zero should not raise an exception, then the same goes for overflow and underflow, which I hope you'd agree should not pass silently in a high-level language like Python.

Reply:
Oh, I see that warning too now.  Looks like numpy only raises it if you're doing stuff in an interactive shell.

RE exceptions: but Python does pass underflow and overflow silently:
```python
>>> 1e-1000
0.0
>>> 1e1000
inf
```

The Zen also says that "practicality beats purity".  Some errors are just too small to fuss with, and Python clearly follows that logic when it comes to the floating-point underflow, overflow, and inexact arithmetic errors.  I think divide-by-zero should also be in that category, partly so that its behavior matches with most other programming languages (I can't think of any other language where divide-by-zero is a halting exception) and partly because that `inf` is often still useful (`math.atan(1 / 0)` is probably the best example).  Oh, well.

Reply:
Overflow can indeed happen:
```Python
>>> 1e100**4
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OverflowError: (34, 'Result too large')
```

Python apparently does some checks at compile-time as your example demonstrates:
```Python
>>> import dis
>>> dis.dis('x = 1e1000')
  1           0 LOAD_CONST               0 (inf)
              2 STORE_NAME               0 (x)
              4 LOAD_CONST               1 (None)
              6 RETURN_VALUE
```
so you see that the interpreter never even sees the value 1e1000. Interestingly, it does appear that underflow [doesn't exist in Python](https://docs.python.org/3/library/exceptions.html#ArithmeticError). While divide-by-zero is not a halting exception in most other languages, most other languages *do* allow you to trap floating point exceptions, which I can personally attest is extremely valuable (when you compile OpenMC with -Ddebug=on, FPEs are trapped). If Python didn't raise exceptions on these conditions, they'd have to provide some other way to trap them. Arguably, a context manager might have been better like [numpy.seterr](https://docs.scipy.org/doc/numpy-1.12.0/reference/generated/numpy.seterr.html#numpy.seterr) (but alas, context managers didn't exist in the early days of Python).